crates/red_knot_python_semantic/src/types.rs

```diff
@@ -782,18 +817,40 @@ impl<'db> Type<'db> {
         // TODO: The following is a workaround that is required to unify the two different versions
         // of `NoneType` and `NoDefaultType` in typeshed. This should not be required anymore once
         // we understand `sys.version_info` branches.
-        self == other
-            || matches!((self, other),
-                (
-                    Type::Instance(InstanceType { class: self_class }),
-                    Type::Instance(InstanceType { class: target_class })
-                )
-                if {
-                    let self_known = self_class.known(db);
-                    matches!(self_known, Some(KnownClass::NoneType | KnownClass::NoDefaultType))
-                        && self_known == target_class.known(db)
-                }
+        match (self, other) {
+            (
+                Type::Instance(InstanceType { class: self_class }),
+                Type::Instance(InstanceType {
+                    class: target_class,
+                }),
+            ) if {
+                let self_known = self_class.known(db);
+                matches!(
+                    self_known,
+                    Some(KnownClass::NoneType | KnownClass::NoDefaultType)
+                ) && self_known == target_class.known(db)
+            } =>
+            {
+                true
+            }
+            (
+                Type::SubclassOf(SubclassOfType {
+                    base: ClassBase::Class(object_class),
+                }),
+                Type::Instance(InstanceType { class: type_class }),
             )
+            | (
+                Type::Instance(InstanceType { class: type_class }),
+                Type::SubclassOf(SubclassOfType {
+                    base: ClassBase::Class(object_class),
+                }),
+            ) if object_class.is_known(db, KnownClass::Object)
+                && type_class.is_known(db, KnownClass::Type) =>
+            {
+                true
+            }
```

Comment:
```suggestion
            ) => {
                object_class.is_known(db, KnownClass::Object)
                    && type_class.is_known(db, KnownClass::Type)
            }
```

Reply:
This one should be safe because a `Type::Instance` can't be `==` to a `Type::SubclassOf`.